core/vm/contracts.go

```diff
@@ -88,47 +101,234 @@ func (c *ecrecover) Run(in []byte) ([]byte, error) {
+
+		input = common.RightPadBytes(input, offset+32)
+		if expLen.Cmp(big.NewInt(32)) > 0 {
+			expHead = new(big.Int).SetBytes(input[offset : offset+32])
+		} else {
+			expHead = new(big.Int).SetBytes(input[offset : offset+int(expLen.Uint64())])
+		}
+	}
+	// Calculate the adjusted exponent length
+	var msb int
+	if bitlen := expHead.BitLen(); bitlen > 0 {
+		msb = bitlen - 1
+	}
+	adjExpLen := new(big.Int)
+	if expLen.Cmp(big.NewInt(32)) > 0 {
+		adjExpLen.Sub(expLen, big.NewInt(32))
+		adjExpLen.Mul(big.NewInt(8), adjExpLen)
+	}
+	adjExpLen.Add(adjExpLen, big.NewInt(int64(msb)))
+
+	// Calculate the gas cost of the operation
+	gas := new(big.Int).Set(math.BigMax(modLen, baseLen))
+	switch {
+	case gas.Cmp(big.NewInt(64)) <= 0:
+		gas.Mul(gas, gas)
+	case gas.Cmp(big.NewInt(1024)) <= 0:
+		gas = new(big.Int).Add(
+			new(big.Int).Div(new(big.Int).Mul(gas, gas), big.NewInt(4)),
+			new(big.Int).Sub(new(big.Int).Mul(big.NewInt(96), gas), big.NewInt(3072)),
+		)
+	default:
+		gas = new(big.Int).Add(
+			new(big.Int).Div(new(big.Int).Mul(gas, gas), big.NewInt(16)),
+			new(big.Int).Sub(new(big.Int).Mul(big.NewInt(480), gas), big.NewInt(199680)),
+		)
+	}
+	gas.Mul(gas, math.BigMax(adjExpLen, big.NewInt(1)))
+	gas.Div(gas, new(big.Int).SetUint64(params.ModExpQuadCoeffDiv))
+
+	if gas.BitLen() > 64 {
+		return math.MaxUint64
+	}
+	return gas.Uint64()
+}
+
+func (c *bigModExp) Run(input []byte) ([]byte, error) {
+	// Pad the input with zeroes to the minimum size to read the field lengths
+	input = common.RightPadBytes(input, 96)
+
+	var (
+		baseLen = new(big.Int).SetBytes(input[:32]).Uint64()
+		expLen  = new(big.Int).SetBytes(input[32:64]).Uint64()
+		modLen  = new(big.Int).SetBytes(input[64:96]).Uint64()
+	)
+	input = input[96:]
+
+	// Pad the input with zeroes to the minimum size to read the field contents
+	input = common.RightPadBytes(input, int(baseLen+expLen+modLen))
+
+	var (
+		base = new(big.Int).SetBytes(input[:baseLen])
+		exp  = new(big.Int).SetBytes(input[baseLen : baseLen+expLen])
+		mod  = new(big.Int).SetBytes(input[baseLen+expLen : baseLen+expLen+modLen])
+	)
+	if mod.BitLen() == 0 {
+		// Modulo 0 is undefined, return zero
+		return common.LeftPadBytes([]byte{}, int(modLen)), nil
+	}
+	return common.LeftPadBytes(base.Exp(base, exp, mod).Bytes(), int(modLen)), nil
+}
+
+// bn256Add implements a native elliptic curve point addition.
+type bn256Add struct{}
+
+// RequiredGas returns the gas required to execute the pre-compiled contract.
+func (c *bn256Add) RequiredGas(input []byte) uint64 {
+	return params.Bn256AddGas
+}
+
+func (c *bn256Add) Run(input []byte) ([]byte, error) {
+	// Ensure we have enough data to operate on
+	input = common.RightPadBytes(input, 128)
+
+	// Parse the coordinates of the curve points
+	x, err := bn256.NewCurvePoint(input[:64])
+	if err != nil {
+		return nil, err
+	}
+	y, err := bn256.NewCurvePoint(input[64:128])
+	if err != nil {
+		return nil, err
+	}
+	// Add the two curve points and return the result
```

Comment:
I know you like to comment things excessively - and usually I just skip over these - but this one triggered me. "add ... and return" followed by `x.Add(...); return ...`. Please remove "XXX and return" comments.